File: SWEA/D2/1974. 스도쿠 검증/스도쿠 검증.py
# 1974. 스도쿠 검증
# 난이도 : D2

T = int(input())

for test_case in range(1,T+1):
    sudoku = [list(map(int,input().split())) for _ in range(9)] # 스도쿠 입력 받기

    # 가로, 세로, 3x3안의 값
    flag_width = False
    flag_length = False
    flag_3x3 = False

    # 들어가야 하는 수
    num_list = [1,2,3,4,5,6,7,8,9]

    # 가로일 경우를 구해보자.
    for row in sudoku:
        temp_row = sorted(row)
        if num_list == temp_row:
            flag_width = True
        else:
            flag_width = False
            break

    # 세로의 경우를 구해보자.
    for i in range(9):
        temp_list = list()
        for j in range(9):
            temp_list.append(sudoku[j][i])
        temp_list.sort()
        if num_list == temp_list:
            flag_length = True
        else:
            flag_length = False
            break

    # 3 x 3 부분들의 합이 모두 45일 경우
    for i in range(0,9,3):
        for j in range(0,9,3):
            sum = 0
            for k in range(i, i+3):
                for l in range(j, j+3):
                    sum += sudoku[k][l]
            if sum == 45:
                flag_3x3 = True
            else:
                flag_3x3 = False
                break
        if sum != 45:
            break

    # 정답이라면 1 출력, 정답이 아닐시 0 출력
    if flag_width and flag_length and flag_3x3:
        answer = 1
    else:
        answer = 0

    print("#{} {}".format(test_case, answer))

File: SWEA/D2/1974. 스도쿠 검증/스도쿠 검증.java
// 1974. 스도쿠 검증
// 난이도 : D2

import java.util.Scanner;
import java.util.Arrays;

public class Solution {
    static int[][] sudoku; // 스도쿠를 static으로 선언
    static int[] checkArr = {1, 2, 3, 4, 5, 6, 7, 8, 9}; // 비교 대상 배열

    public static void main(String args[]) throws Exception {
        Scanner scanner = new Scanner(System.in);

        int T = scanner.nextInt(); // 테스트 케이스 갯수 입력 받기

        // 9 x 9 스도쿠를 T개 입력 받는다.
        for (int test_case = 1; test_case <= T; test_case++) {

            // 9x9 이차원 배열에 스도쿠를 입력 받자.
            sudoku = new int[9][9];
            for (int i = 0; i < 9; i++) {
                for (int j = 0; j < 9; j++) {
                    sudoku[i][j] = scanner.nextInt();
                }
            }

            // 이제 그 스도쿠가 맞는지 검증해 보자.
            // 스도쿠는 가로, 세로, 그리고 3x3 공간에 1,2,3,4,5,6,7,8,9 9개의 숫자가 존재해야 한다.
            boolean flag_width = checkWidth();
            boolean flag_height = checkHeight();
            boolean flag_3x3 = check3x3();

            int answer;

            if (flag_width && flag_height && flag_3x3) {
                answer = 1;
            } else {
                answer = 0;
            }


            // 1이면 정답, 0이면 오답
            System.out.println("#" + test_case + " " + answer);
        }
    }

    // 가로를 체크한다.
    static boolean checkWidth() {
        // 가로줄에 있는 것들을 체크
        for (int i = 0; i < 9; i++) {
            int[] temp = new int[9]; // 가로줄을 담을 임시 배열
            //  i번째 행의 sudoku 값을 temp 배열에 담아준다.
            for (int j = 0; j < 9; j++) {
                temp[j] = sudoku[i][j];
            }
            Arrays.sort(temp); // 정렬

            // checkArr의 원소들과 비교하여 틀릴경우 바로 false를 반환한다.
            for (int j = 0; j < 9; j++) {
                if (temp[j] != checkArr[j])
                    return false;
            }

        }
        // 다 통과하고 다 성립하면 true 반환
        return true;
    }

    static boolean checkHeight() {
        // 세로줄에 있는 것
        for (int i = 0; i < 9; i++) {
            int[] temp = new int[9]; // 세로줄을 담을 임시 배열

            //  j번째 열의 sudoku 값을 temp 배열에 담아준다.
            for (int j = 0; j < 9; j++) {
                temp[j] = sudoku[j][i];
            }

            Arrays.sort(temp); // 정렬

            // checkArr의 원소들과 비교하여 틀릴경우 바로 false를 반환한다.
            for (int j = 0; j < 9; j++) {
                if (temp[j] != checkArr[j])
                    return false;
            }

        }
        // 다 통과하고 다 성립하면 true 반환
        return true;
    }

    static boolean check3x3() {
        // 3x3 부분의 원소들을 담아야한다.
        // [0][0] ~ [2][2], [0][3] ~ [2][5], [0][6] ~ [2][8]
        // [3][0] ~ [5][2], [3][3] ~ [5][5], [3][6] ~ [5][8]
        // [6][0] ~  [8][2], [6][3] ~ [8][5], [6][6] ~ [8][8]
        for (int i = 0; i < 3; i++){
            for (int j = 0; j < 3; j++){
                int[] temp = new int[9];
                int index = 0;

                // 4중 반복문?!?!?!
                for (int k = 0; k < 3; k++){
                    for (int l = 0; l < 3; l++){
                        temp[index++] = sudoku[i * 3 + k][j * 3 + l];
                    }
                }

                Arrays.sort(temp); // 정렬

                // checkArr의 원소들과 비교하여 틀릴경우 바로 false를 반환한다.
                for (int k = 0; k < 9; k++) {
                    if (temp[k] != checkArr[k])
                        return false;
                }
            }
        }
        // 다 통과하고 다 성립하면 true 반환
        return true;
    }
}